### PR TITLE
Convert CM to TS

### PR DESCRIPTION
**Pull request by jhoward1994**
### What does it do?

Converts the content manager to TS

### CODE REVIEW

**Comment by jhoward1994** [comment]
@Marc-Roig this branch currently has a basic conversion of the CM to typescript, using `any` pretty much everywhere and pack up to build the package.

I'm trying to remember what we spoke about in terms of getting this merged to main? I know we want to avoid a PR like this staying open for ages like https://github.com/strapi/strapi/pull/18232

Should the plan be to make all the tests pass asap here, merge the branch and then work on specific types for different parts of the CM BE?

**Comment by joshuaellis** [comment]
Would it not make sense to convert a section at a time and then merge that in? You could temporarily set "allowJS" to be true on your tsconfig which might make the import errors go away

**Comment by Marc-Roig** [comment]
@joshuaellis we could try to do so, just that in the admin package we found that tot be very difficult. The import hierarchy was not very clear and we found that for some files we needed to type other sections. So in the end we found it easier just to type everything first and incrementally add the types later.

**Comment by Marc-Roig** [comment]
>Should the plan be to make all the tests pass asap here, merge the branch and then work on specific types for different parts of the CM BE?

In my opinion, yes, the sooner we merge things the better for people.

I would warn people working on the CM server before we merge this though.

File: packages/core/admin/_internal/node/createBuildContext.ts [hunk 1]
```diff
@@ -218,7 +218,11 @@ const filterPluginsByAdminEntry = (plugin: Plugin) => {
 
     return isLocalPluginWithLegacyAdminFile;
   } catch (err) {
-    if (isError(err) && 'code' in err && err.code === 'MODULE_NOT_FOUND') {
+    if (
+      isError(err) &&
+      'code' in err &&
+      (err.code === 'MODULE_NOT_FOUND' || err.code === 'ERR_PACKAGE_PATH_NOT_EXPORTED')
```
Comment (Marc-Roig) [inline packages/core/admin/_internal/node/createBuildContext.ts:224]: why was this change needed?

File: packages/core/content-manager/server/src/controllers/collection-types.ts [hunk 2]
```diff
@@ -22,6 +21,7 @@ module.exports = {
 
     const permissionQuery = await permissionChecker.sanitizedQuery.read(query);
 
+    // @ts-ignore TODO
```
Comment (Marc-Roig) [inline packages/core/content-manager/server/src/controllers/collection-types.ts:24]: In this case probably it's not even important, but I would normally use ts-expect-error instead of ts-ignore. While it can be more annoying somethimes , if we leave the message there and the original ts errors disappears, it will warn us

File: packages/core/admin/_internal/node/createBuildContext.ts (hunk 1 above)
Comment (joshuaellis) [reply packages/core/admin/_internal/node/createBuildContext.ts:224]: I made this change, because the we get an error about `strapi-admin` missing from the exports package, which realistically we can ignore because not all plugins have an admin part :)

**Comment by joshuaellis** [comment]
> @joshuaellis we could try to do so, just that in the admin package we found that tot be very difficult. The import hierarchy was not very clear and we found that for some files we needed to type other sections. So in the end we found it easier just to type everything first and incrementally add the types later.

Whatever works for you guys, i think we're in agreement we want to avoid the admin package PR size again lol

**Comment by Marc-Roig** [comment]
yes, that was too much. This only adds ts files and a bunch of any's, so the next prs should be more l ight weight

File: packages/core/content-manager/server/src/services/permission-checker.ts [hunk 3]
```diff
@@ -12,76 +10,81 @@ const ACTIONS = {
...
+      // @ts-expect-error TODO
+      sanitizedQuery[action] = (query: any) => sanitizedQuery(query, ACTIONS[action]);
     });
 
     // Permission utils shortcuts
     Object.keys(ACTIONS).forEach((action) => {
-      can[action] = (...args) => can(ACTIONS[action], ...args);
-      cannot[action] = (...args) => cannot(ACTIONS[action], ...args);
+      // @ts-expect-error TODO
+      can[action] = (...args: any) => can(ACTIONS[action], ...args);
```
Comment (jhoward1994) [inline packages/core/content-manager/server/src/services/permission-checker.ts:85]: @Marc-Roig this confuses me a bit

the ts error is `type 'string' can't be used to index type '(action: any, entity: any, field: any) => any` which makes sense to me as the only can defined here is the func at L23. The same problem seems to exist before this TS change though

Comment (Marc-Roig) [reply packages/core/content-manager/server/src/services/permission-checker.ts:85]: maybe this refers to the "action" which is maybe not interpreted as a string, can you try to define the ACTIONS object as :

```ts
const ACTIONS = {
  read: 'plugin::content-manager.explorer.read',
  create: 'plugin::content-manager.explorer.create',
  update: 'plugin::content-manager.explorer.update',
  ...
} as const;
```

Comment (Marc-Roig) [reply packages/core/content-manager/server/src/services/permission-checker.ts:85]: ah 💡 missunderstood what was happening. Yes, I dont know if this can actually be properly typed 🤔  maybe @Convly  knows more about this.

These are shortcuts aadded on top of what originally is a function, javascript allows this sourcery and probably typescript has some magic too. Lets leave it as a ts epxect error for now if it's not easy to fix

**Review by joshuaellis — APPROVED** [review]
@Marc-Roig should most likely be the sign-off reviewer, but the changes to admin look fine 👍🏼

**Review by Marc-Roig — APPROVED** [review]

**Comment by joshuaellis** [comment]
E2E tests are broken 🤔 not fully looked as to why.

**Comment by jhoward1994** [comment]
Yeah I've been trying to look into it but havent cracked it yet

image


It looks like a call to `formatMessage` within the CM isn't being given the right shape object?

**Comment by jhoward1994** [comment]
Merging this PR. The failing FE tests are unrelated to these changes and is an issue being fixed elsewhere https://github.com/strapi/strapi/pull/18794

I've verified that the tests are passing locally, done a QA for the CM and checked that the BE routes are working
